gflownet.py

```diff
@@ -487,25 +490,92 @@ def __init__(
     def parameters(self):
         return self.model.parameters()
 
-    def backward_sample(self, env, policy="model", done=False):
+    def forward_sample(self, envs, times, policy="model", model=None, temperature=1.0):
+        """
+        Performs a forward action on each environment of a list.
+
+        Args
+        ----
+        env : list of GFlowNetEnv or derived
+            A list of instances of the environment
+
+        times : dict
+            Dictionary to store times
+
+        policy : string
+            - model: uses self.model to obtain the sampling probabilities.
+            - uniform: samples uniformly from the action space.
+
+        model : torch model
+            Model to use as policy if policy="model"
+
+        temperature : float
+            Temperature to adjust the logits by logits /= temperature
+        """
+        if not isinstance(envs, list):
+            envs = [envs]
+        states = [env.state2obs() for env in envs]
+        mask_invalid_actions = [env.get_mask_invalid_actions() for env in envs]
+        random_action = self.rng.uniform()
+        t0_a_model = time.time()
+        if policy == "model":
+            with torch.no_grad():
```

Comment:
do we have to specify model.train(mode = False), to remove the dropout option when sampling with the model ?

Reply:
Uhmm I am not sure if `no_grad()` does that too, but we are not using dropout anyway, do we?

Reply:
I thought that no_grad() only disabled the heavy computation of gradients. 
Oh yes, I mixed up with another MLP that I wrote in the same spirit as the current proxy for the GFN, and where I added dropout for the training. :)
Current implementation is indeed  calling Sequential built-in MLP !